stdlib/asm/crypto/hashes/rpo.masm

```diff
@@ -100,3 +100,155 @@ export.hash_memory
+
+        # get the flag whether the drop counter is equal 1
+        dup.14 eq.1
+        # => [drop_counter == 1, latch = 0, e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # update the latch: if drop_counter == 1, latch will become 1
+        or
+        # => [latch', e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
+
+        # save the latch value 
+        dup movdn.14
+        # => [latch', e_1, 0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', latch', drop_counter]
+
+        # if latch == 1, drop 0; otherwise drop e_1
+        cdrop
+        # => [e_1_or_0, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', latch', drop_counter]
+
+        # move the calculated value down the stack 
+        movdn.6
+        # => [e_2, e_3, d_0, d_1, d_2, 0, e_1_or_0, d_3, A', latch', drop_counter]
+
+        ### 2nd value ########################################################
+
+        # repeat the above process but now compare drop_counter to 2
+        push.0 swap
+        movup.13 dup.14 eq.2 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [e_3, d_0, d_1, d_2, 0, e_1_or_0, e_2_or_0, d_3, A', latch', drop_counter]
+
+        ### 3rd value ########################################################
+
+        # repeat the above process but now compare drop_counter to 3
+        push.0 swap
+        movup.13 dup.14 eq.3 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [d_0, d_1, d_2, 0, e_1_or_0, e_2_or_0, e_3_or_0, d_3, A', latch', drop_counter]
+
+        ### 4th value ########################################################
+
+        # repeat the above process but now compare drop_counter to 4
+        push.0 swap
+        movup.13 dup.14 eq.4 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [d_1, d_2, 0, e_1_or_0, e_2_or_0, e_3_or_0, d_0_or_0, d_3, A', latch', drop_counter]
+
+        ### 5th value ########################################################
+
+        # repeat the above process but now compare drop_counter to 5
+        push.0 swap
+        movup.13 dup.14 eq.5 or
+        dup movdn.14
+        cdrop movdn.6
+        # => [d_2, 0, e_1_or_0, e_2_or_0, e_3_or_0, d_0_or_0, d_1_or_0, d_3, A', latch', drop_counter]
+
+        ### 6th value ########################################################
+
+        # repeat the above process but now compare drop_counter to 6
+        push.0 swap
+        movup.13 movup.14 eq.6 or
+        cdrop movdn.6
+        # => [0, e_1_or_0, e_2_or_0, e_3_or_0, d_0_or_0, d_1_or_0, d_2_or_0, d_3, A']
+        # or in other words
+        # => [C, B, A', ... ]
+        # notice that we don't need to check the d_3 value: entering the if.true branch means that 
+        # we have number of elements not divisible by 8, so we will have at least one element to 
+        # hash here (which turns out to be d_3)
+
+        hperm
+        # => [F, E, D]
+
+        exec.squeeze_digest
+        # => [E]
+    else
+        dropw movdnw.2 dropw dropw
```

Comment:
Does this not drop an extra element? (based on line 133 there should be only 15 elements on the stack).

Also, can we incorporate `exec.squeeze_digest` into this (even if it takes an extra cycle).